File: letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        size = len(digits)
        keyPad = {
            '2':"abc",
            '3':"def",
            '4':"ghi",
            '5':'jkl',
            '6':'mno',
            '7':'pqrs',
            '8':'tuv',
            '9':'wxyz'
        }
        res = []
        def getCombinations(idx,temp):
            if idx>=size:
                if temp!='':
                    res.append(temp)
                return
            
            for i in range(len(keyPad[digits[idx]])):
                getCombinations(idx+1,temp+keyPad[digits[idx]][i])
            return
        
        getCombinations(0,'')
        return res

File: letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.java
class Solution {
    
    public static void recurCombination(String digits, String[] phoneNumber, int idx, String curr, List<String> resCombination){
        if(idx>=digits.length()){
            if(curr.equals("") == false){
                resCombination.add(curr);    
            }
            return;
        }
        int idxVal = digits.charAt(idx) - '0' - 2;
        for(int i=0; i<phoneNumber[idxVal].length();i++){
            recurCombination(digits,phoneNumber,idx+1,curr+phoneNumber[idxVal].charAt(i),resCombination);
        }
        return;
    }
    
    public List<String> letterCombinations(String digits) {
        
        String[] phoneNumber = {
            "abc",
            "def",
            "ghi",
            "jkl",
            "mno",
            "pqrs",
            "tuv",
            "wxyz"
        };
        ArrayList<String> resCombination = new ArrayList<>();
        //System.out.println(phoneNumber[0]);
        recurCombination(digits,phoneNumber,0,"",resCombination);
        
        return resCombination;
        
        
    }
}